pkg/splunk/enterprise/util.go

```diff
@@ -1939,3 +1939,92 @@ func migrateAfwFromPhase2ToPhase3(ctx context.Context, client splcommon.Controll
+	reqLogger := log.FromContext(ctx)
+	scopedLog := reqLogger.WithName("updateCRStatus").WithValues("name", dirtyCR.GetName(), "namespace", dirtyCR.GetNamespace())
+
+	var retryCnt int
+	for retryCnt = 0; retryCnt < maxRetryCountForCRStatusUpdate; retryCnt++ {
+		cr, err := fetchCurrentCRWithStatusUpdate(ctx, client, dirtyCR)
+		if err != nil {
+			scopedLog.Error(err, "Unable to Read the latest CR from the K8s")
+			continue
+		}
+
+		err = client.Status().Update(ctx, cr)
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Duration(retryCnt) * 10 * time.Millisecond)
+	}
+
+	if retryCnt >= maxRetryCountForCRStatusUpdate {
+		scopedLog.Error(nil, "Status update failed", "cr version", dirtyCR.GetResourceVersion(), "reTry count", retryCnt)
+	}
+}
+
+// fetchCurrentCRWithStatusUpdate returns a CR (fresh Read) with latest status copied
+func fetchCurrentCRWithStatusUpdate(ctx context.Context, client splcommon.ControllerClient, dirtyCR splcommon.MetaObject) (splcommon.MetaObject, error) {
+	namespacedName := types.NamespacedName{Name: dirtyCR.GetName(), Namespace: dirtyCR.GetNamespace()}
+
+	var err error
+	switch dirtyCR.GetObjectKind().GroupVersionKind().Kind {
+	case "Standalone":
+		latestStdlnCR := &enterpriseApi.Standalone{}
+		err = client.Get(ctx, namespacedName, latestStdlnCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.Standalone).Status.DeepCopyInto(&latestStdlnCR.Status)
+		return latestStdlnCR, nil
+
+	case "LicenseMaster":
+		latestLmCR := &enterpriseApi.LicenseMaster{}
+		err = client.Get(ctx, namespacedName, latestLmCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.LicenseMaster).Status.DeepCopyInto(&latestLmCR.Status)
+		return latestLmCR, nil
+
+	case "SearchHeadCluster":
+		latestShcCR := &enterpriseApi.SearchHeadCluster{}
+		err = client.Get(ctx, namespacedName, latestShcCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.SearchHeadCluster).Status.DeepCopyInto(&latestShcCR.Status)
+		return latestShcCR, nil
+
+	case "IndexerCluster":
+		latestIdxcCR := &enterpriseApi.IndexerCluster{}
+		err = client.Get(ctx, namespacedName, latestIdxcCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.IndexerCluster).Status.DeepCopyInto(&latestIdxcCR.Status)
+		return latestIdxcCR, nil
+
+	case "ClusterMaster":
+		latestCmCR := &enterpriseApi.ClusterMaster{}
+		err = client.Get(ctx, namespacedName, latestCmCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.ClusterMaster).Status.DeepCopyInto(&latestCmCR.Status)
+		return latestCmCR, nil
+
+	case "MonitoringConsole":
+		latestMcCR := &enterpriseApi.MonitoringConsole{}
+		err = client.Get(ctx, namespacedName, latestMcCR)
+		if err != nil {
+			return nil, err
+		}
+		dirtyCR.(*enterpriseApi.MonitoringConsole).Status.DeepCopyInto(&latestMcCR.Status)
+		return latestMcCR, nil
+	}
```

Comment:
Can we restructure it like this? -
```
var latestCR splcommon.MetaObject
var status client.StatusWriter
switch dirtyCR.GetObjectKind().GroupVersionKind().Kind {
        
	case "Standalone":
		latestCR = &enterpriseApi.Standalone{}
		status = dirtyCR.(*enterpriseApi.Standalone).Status
	case "LicenseMaster":
		latestCR := &enterpriseApi.LicenseMaster{}
		status = dirtyCR.(*enterpriseApi.LicenseMaster).Status

	case "SearchHeadCluster":
		latestCR := &enterpriseApi.SearchHeadCluster{}
		status = dirtyCR.(*enterpriseApi.SearchHeadCluster).Status

	case "IndexerCluster":
		latestCR := &enterpriseApi.IndexerCluster{}
		status  = dirtyCR.(*enterpriseApi.IndexerCluster).Status

	case "ClusterMaster":
		latestCR := &enterpriseApi.ClusterMaster{}
		status = dirtyCR.(*enterpriseApi.ClusterMaster).Status

	case "MonitoringConsole":
		latestCR := &enterpriseApi.MonitoringConsole{}
		status = dirtyCR.(*enterpriseApi.MonitoringConsole).Status
	}

err = client.Get(ctx, namespacedName, latestCR)
if err != nil {
	  return nil, err
}
status.DeepCopyInto(&latestCR.Status)
return latestCR, nil
```

Reply:
Here, we are just using the auto generated method DeepCopyInto(), so, we can't use the interface.

Reply:
Yeah probably you are right. But my motive was to reduce the duplicate code as much as possible, not sure if there is any better way though.

Reply:
Agree.  Right now, as part of the API, all we have is the type declarations, and some auto generated deep copy code. Since the receiver needs to be in the same package,  we can try to add an interface (that wraps the Status) in the V3 API itself and use it. Probably, we need to introduce a new file and try. Again, it needs some thought and re-work, as we need to pass the client as well, which is from a different package.